File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-java/src/main/java/com/opspilot/team/TeamGraph.java
package com.opspilot.team;

import com.opspilot.domain.TraceEvent;

import java.util.ArrayList;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;

/**
 * Porte de team-graph.ts: a mesma maquina de estado (sem depender do
 * StateGraph do LangGraph) -- supervisor decide o proximo papel, cada papel
 * contribui uma entrada no blackboard, ate a decisao "done" ou o teto de
 * handoffs (MAX_HANDOFFS).
 */
public final class TeamGraph {

    public static final int MAX_HANDOFFS = 8;
    public static final String CAP_REACHED_PREFIX = "teto de handoffs atingido";
    public static final String INVALID_DECISION_PREFIX = "decisao invalida do supervisor";

    private final DecideNextFn decideNext;
    private final Map<TeamRole, RoleRunner> roleRunners;
    private final int supervisorLlmCalls;

    public TeamGraph(DecideNextFn decideNext, Map<TeamRole, RoleRunner> roleRunners, int supervisorLlmCalls) {
        this.decideNext = decideNext;
        this.roleRunners = new EnumMap<>(roleRunners);
        this.supervisorLlmCalls = supervisorLlmCalls;
    }

    public record Result(String answer, List<TraceEvent> trace, int llmCalls) {
    }

    public Result run(String message) {
        List<BlackboardEntry> blackboard = new ArrayList<>();
        List<TraceEvent> trace = new ArrayList<>();
        int handoffCount = 0;
        int llmCalls = 0;
        String lastBrief = "";

        while (true) {
            if (handoffCount >= MAX_HANDOFFS) {
                String content = CAP_REACHED_PREFIX + ": encerrando com o conteudo do blackboard";
                trace.add(TraceEvent.handoff("supervisor", "done", content));
                lastBrief = "";
                break;
            }

            SupervisorDecision decision;
            try {
                decision = decideNext.decideNext(message, blackboard, handoffCount);
            } catch (RuntimeException e) {
                String content = INVALID_DECISION_PREFIX + ": " + e.getMessage();
                trace.add(TraceEvent.handoff("supervisor", "done", content));
                llmCalls += supervisorLlmCalls;
                lastBrief = "";
                break;
            }

            trace.add(TraceEvent.handoff(
                    "supervisor", decision.isDone() ? "done" : decision.next().name().toLowerCase(), decision.brief()));
            llmCalls += supervisorLlmCalls;

            if (decision.isDone()) {
                lastBrief = decision.brief();
                break;
            }

            handoffCount++;
            RoleRunner runner = roleRunners.get(decision.next());
            try {
                RoleRunResult result = runner.run(new RoleRunInput(message, decision.brief(), List.copyOf(blackboard)));
                blackboard.add(result.entry());
                trace.addAll(result.trace());
                llmCalls += result.llmCalls();
            } catch (RuntimeException e) {
                String content = "erro no papel " + decision.next().name().toLowerCase() + ": " + e.getMessage();
                blackboard.add(new BlackboardEntry(decision.next(), BlackboardEntry.Kind.ERROR, decision.brief(), content));
                trace.add(TraceEvent.observation(decision.next().name().toLowerCase(), content));
            }
        }

        String answer;
        String trimmedBrief = lastBrief.trim();
        if (!trimmedBrief.isEmpty()) {
            answer = trimmedBrief;
        } else if (!blackboard.isEmpty()) {
            answer = "Resumo do blackboard:\n" + Blackboard.render(blackboard);
        } else {
            answer = "A equipe encerrou sem contribuicoes no blackboard para: " + message;
        }
        trace.add(TraceEvent.answer("supervisor", answer));

        return new Result(answer, trace, llmCalls);
    }
}

File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-go/team/team_graph.go
package team

import (
	"fmt"
	"strings"

	"opspilot/domain"
)

// MaxHandoffs é o porte de MAX_HANDOFFS (team/team-graph.ts) -- teto de
// delegações do supervisor por turno (spec 018, US5).
const MaxHandoffs = 8

// Prefixos estáveis, testados e documentados em contracts/trace-handoff.md
// no original.
const (
	CapReachedPrefix      = "teto de handoffs atingido"
	InvalidDecisionPrefix = "decisao invalida do supervisor"
)

// TeamGraph é o porte de team-graph.ts: a mesma máquina de estado (sem
// depender do StateGraph do LangGraph, que não tem equivalente maduro em Go)
// -- supervisor decide o próximo papel, cada papel contribui uma entrada no
// blackboard, até a decisão "done" ou o teto de handoffs (MaxHandoffs). O
// estado da rodada (blackboard, trace, contadores) vive em variáveis locais
// de Run; não há grafo declarativo, só a função de transição explícita.
type TeamGraph struct {
	DecideNext         DecideNextFn
	RoleRunners        map[Role]RoleRunner
	SupervisorLLMCalls int
}

// NewTeamGraph cria o coordenador do time. supervisorLLMCalls é quantas
// chamadas de LLM contar por decisão real do supervisor (0 para fakes
// injetados em teste).
func NewTeamGraph(decideNext DecideNextFn, roleRunners map[Role]RoleRunner, supervisorLLMCalls int) *TeamGraph {
	runners := make(map[Role]RoleRunner, len(roleRunners))
	for role, runner := range roleRunners {
		runners[role] = runner
	}
	return &TeamGraph{DecideNext: decideNext, RoleRunners: runners, SupervisorLLMCalls: supervisorLLMCalls}
}

// TeamGraphResult é o porte de TeamGraph.Result (porte Java).
type TeamGraphResult struct {
	Answer   string
	Trace    []domain.TraceEvent
	LLMCalls int
}

// Run é a função coordenadora que orquestra supervisor + papéis até a
// decisão "done" ou o teto de handoffs -- porte comportamental 1:1 de
// runTeamGraph / supervisorNode / roleNode / doneNode (team-graph.ts).
func (g *TeamGraph) Run(message string) TeamGraphResult {
	var blackboard []BlackboardEntry
	var trace []domain.TraceEvent
	handoffCount := 0
	llmCalls := 0
	lastBrief := ""

	for {
		if handoffCount >= MaxHandoffs {
			content := CapReachedPrefix + ": encerrando com o conteudo do blackboard"
			trace = append(trace, domain.Handoff("supervisor", "done", content))
			lastBrief = ""
			break
		}

		decision, err := g.DecideNext(message, blackboard, handoffCount)
		if err != nil {
			content := InvalidDecisionPrefix + ": " + err.Error()
			trace = append(trace, domain.Handoff("supervisor", "done", content))
			llmCalls += g.SupervisorLLMCalls
			lastBrief = ""
			break
		}

		nextLabel := "done"
		if !decision.Done {
			nextLabel = string(decision.Next)
		}
		trace = append(trace, domain.Handoff("supervisor", nextLabel, decision.Brief))
		llmCalls += g.SupervisorLLMCalls

		if decision.Done {
			lastBrief = decision.Brief
			break
		}

		handoffCount++
		runner, ok := g.RoleRunners[decision.Next]
		if !ok {
			content := fmt.Sprintf("erro no papel %s: nenhum role runner configurado", decision.Next)
			blackboard = append(blackboard, BlackboardEntry{Role: decision.Next, Kind: KindError, Brief: decision.Brief, Content: content})
			trace = append(trace, domain.Observation(string(decision.Next), content))
			continue
		}

		blackboardSnapshot := append([]BlackboardEntry(nil), blackboard...)
		result, err := runner.Run(RoleRunInput{Message: message, Brief: decision.Brief, Blackboard: blackboardSnapshot})
		if err != nil {
			content := fmt.Sprintf("erro no papel %s: %s", decision.Next, err.Error())
			blackboard = append(blackboard, BlackboardEntry{Role: decision.Next, Kind: KindError, Brief: decision.Brief, Content: content})
			trace = append(trace, domain.Observation(string(decision.Next), content))
			continue
		}

		blackboard = append(blackboard, result.Entry)
		trace = append(trace, result.Trace...)
		llmCalls += result.LLMCalls
	}

	var answer string
	trimmedBrief := strings.TrimSpace(lastBrief)
	switch {
	case trimmedBrief != "":
		answer = trimmedBrief
	case len(blackboard) > 0:
		answer = "Resumo do blackboard:\n" + RenderBlackboard(blackboard)
	default:
		answer = "A equipe encerrou sem contribuicoes no blackboard para: " + message
	}
	trace = append(trace, domain.Answer("supervisor", answer))

	return TeamGraphResult{Answer: answer, Trace: trace, LLMCalls: llmCalls}
}
